File: Dictionary Class in C/Dictionary.c
// //Alston Huang
// //1471706
// //CMPS012b
// //Structure and functions for DictionaryClient
// //Dictionary.c

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"

//private types

//NodeObj
typedef struct NodeObj{
	char* key;
	char* value;
	struct NodeObj* next;
} NodeObj;

//Node
typedef NodeObj* Node;

//newNode()
//constructor of the Node type
Node newNode(char* k, char* v){
	Node N = malloc(sizeof(NodeObj));
	assert(N!=NULL);
	N->key = k;
	N->value = v;
	N->next = NULL;
	return(N);
}

//freeNode();
//destructor for the Node type
void freeNode(Node* pN){
	if( pN != NULL && *pN != NULL ){
		free(*pN);
		*pN = NULL;
	}
}

//DictionaryObj
typedef struct DictionaryObj{
	Node head;
	Node tail;
	int numItems;
} DictionaryObj;

//constructor for Dictioary type
Dictionary newDictionary(void){
	Dictionary D = malloc(sizeof(DictionaryObj));
	assert(D != NULL );
	D->head = NULL;
	D->tail = NULL;
	D->numItems = 0;
	return(D);
}

//destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
	if( pD != NULL && *pD != NULL){
		if( !isEmpty(*pD) )
			makeEmpty(*pD);
		free(*pD);
		*pD = NULL;
	}
}

//Private Function

//findKey()
Node findKey(Node N, char* k){

	if( N->next == NULL && N->key == k )
		return N;
	
	while( N != NULL ){
		if( N->next->key != k )
			N = N->next;
		else 
			return N;
	}
	return NULL;
}


//ADT Operations

//isEmpty() function
// 1 = true, 0 = false
int isEmpty(Dictionary D){
	if( D == NULL ){
		fprintf(stderr, "Dictionary Error: calling isEmpty() on null Dictionary reference\n");
		exit(EXIT_FAILURE);
	}

	//returns false
	if( D->numItems > 0 )
		return 0;

	//only evaluated if the if statement above fails.
	return 1;
}

//size() returns size of dictionary
int size(Dictionary D){
	return D->numItems;
}


//returns the value v such that (k, v) is in D, or returns NULL if not
char* lookup(Dictionary D, char* k){
	Node N = D->head;

	if( N == NULL ){
		fprintf(stderr, "Dictionary Error: calling lookup() on NULL Dictionary\n");
		exit(EXIT_FAILURE);
	}

	while( N != NULL ){
		if( strcmp(N->key, k) == 0 ){
			return N->value;
		}
		N = N->next;
	}
	
	return NULL;
}

//insert()
//pre condition: lookup(D, k) == NULL
void insert(Dictionary D, char* k, char* v){

	Node temp = newNode(k, v);

	//special case for empty dictionary
	if( D->numItems == 0 ){
		D->head = D->tail = temp;
	}

	else if( lookup(D, k) != NULL ){
		fprintf(stderr, "Dictionary Error: calling insert() on Duplicate Key\n");
		exit(EXIT_FAILURE);
	}

	else{
		D->tail->next = temp;
		D->tail = temp;
	}

	D->numItems++;

}

void delete(Dictionary D, char* k){
	if( lookup(D, k) == NULL ){
		fprintf(stderr, "Dictionary Error: calling delete() on Non-Existent Key\n");
		exit(EXIT_FAILURE);
	}

	Node N = D->head;

	//special case one element in dictionary
	if( D->numItems <= 1 ){
		D->head = NULL;
	}

	//special case if first node in dictionary contains the key
	else if( strcmp(N->key, k) == 0 ){
		D->head = N->next;
	}

	else{
		while( strcmp(N->next->key, k) != 0 ){
			N = N->next;
		}
		N->next = N->next->next;
	}
	
	D->numItems--;
}


void makeEmpty(Dictionary D){
	D->head = NULL;
	D->tail = NULL;
	D->numItems = 0;
}

void printDictionary(FILE* out, Dictionary D){

	if( D == NULL ){
		fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL Dictionary\n");
	}

	for( Node N = D->head; N != NULL; N = N->next ){
		fprintf(out, "%s %s \n", N->key, N->value );
	}
}

File: Dictionary_LinkedList_Java/Dictionary.java
public class Dictionary implements DictionaryInterface{

	//private inner Node class
	private class Node{
		String value;
		String key;
		Node next;

		Node(String k, String v){
			key = k;
			value = v;
			next = null;
		}
	}

	//Fields for Dictionary class
	private Node head;
	private int numItems;

	//constructor for Dictionary class
	public Dictionary(){
		head = null;
		numItems = 0;
	}

	//private Helper function
	private Node findKey(String k){
		Node N = head;
		while( N != null ){
			if( N.key != k )
				N = N.next;
			else
				return N;
		}
		return null;
	}

	//ADT operations

	//isEmpty()
	public boolean isEmpty(){
		return (numItems == 0);
	}

	//size();
	public int size(){
		return numItems;
	}

	//lookup(String key);
	public String lookup(String k){
		Node N = head;
		while( N != null ){
			if( N.key.equals(k) )
				return N.value;
			else
				N = N.next;
		}
		return null;
	}

	//insert(String k, String v) inserts a new node at the end
	//pre-condition: Dictionary does not currently contain argument key
	public void insert(String k, String v) throws DuplicateKeyException{
		Node N = head;
		Node temp = new Node(k, v);

		//determine if there isn't already a key in this node
		if( lookup(k) != null ){
			throw new DuplicateKeyException("Cannot insert duplicate keys.");
		}

		//make special case for empty dictionary
		if( head == null ){
			head = temp;
		}

		else{
			//run to the node before the last one.
			while( N != null ){
				if( N.next == null )
					break;
				N = N.next;
			}

			N.next = temp;
		}

		numItems++;
	}

	//delete(String key)
	//pre-conditon: Dictionary must contain the argument key
	public void delete(String k) throws KeyNotFoundException{
		Node N = head;

		//check if such key exists, if not throw an exception
		if( lookup(k) == null )
			throw new KeyNotFoundException("Cannot delete non-existent key.");

		else {
			//special case for one node. deletes only if there is one node
			if( numItems <= 1 ){
				head = head.next;
				N.next = null;
			}
			else if( N.key.equals(k) ){
				head = N.next;
			}
			//else, run to the node before key node.
			else{
				while( !N.next.key.equals(k) )
					N = N.next;
				N.next = N.next.next;
			}
		}
		numItems--;
	}

	//makeEmpty()
	public void makeEmpty(){
		head = null;
		numItems = 0;
	}

	//toString()
	public String toString(){
		String string = "";
		Node N = head;

		while( N != null ){
			string = string + N.key + " " + N.value + "\n";
			N = N.next;
		}
		return string;
	}
}